### ISSUE TITLE
Discrepancy between the paper and the code?

### ISSUE THREAD

**Issue by Darel13712** [raminnakhli/Decoupled-Contrastive-Learning#14]
The paper defines negative similarity part of the loss like this:

$$
U_{i,k} = \sum_{\ell \in \{1, 2\}, j \in [1, N], j \neq i} \exp\left(\frac{\langle z_i^{(k)}, z_j^{(\ell)} \rangle}{\tau}\right)
$$

Whereas the code like this:

```python
neg_similarity = torch.cat((torch.mm(z1, z1.t()), cross_view_distance), dim=1) / self.temperature
neg_mask = torch.eye(z1.size(0), device=z1.device).repeat(1, 2)
negative_loss = torch.logsumexp(neg_similarity + neg_mask * SMALL_NUM, dim=1, keepdim=False)
```

As hard as I look at the code, I do not understand what the `+ neg_mask * SMALL_NUM` part does. It basically ads 0 to the diagonals but does not remove them from the calculation, am I right?

If we want to assure that the negative part of the loss does indeed follow the $i \neq j$ and the diagonals are removed we should change
`negative_loss = torch.logsumexp(neg_similarity + neg_mask * SMALL_NUM, dim=1, keepdim=False)`
to
`negative_loss = torch.logsumexp(neg_similarity * (1 - neg_mask + neg_mask * SMALL_NUM), dim=1, keepdim=False)`

Can you please verify this? Am I missing something or is there a mistake in the current implementation of the loss? If so, is this a critical thing or it does not really matter?

**Comment by DeanTM** [comment]
I'm wondering the same thing, and came here looking for answers.

My understanding is that the negative loss should be something like 
```python
negative_loss = torch.logsumexp(neg_similarity.masked_fill(neg_mask, -torch.inf) + SMALL_NUM, dim=1, keepdim=False)
```

so that the diagonals of the `neg_similarity` are zeroed-out in the exp-part of logsumexp.
